lightning/src/ln/channelmanager.rs

```diff
@@ -1680,30 +1682,47 @@ macro_rules! handle_new_monitor_update {
 
 macro_rules! process_events_body {
 	($self: expr, $event_to_handle: expr, $handle_event: expr) => {
-		// We'll acquire our total consistency lock until the returned future completes so that
-		// we can be sure no other persists happen while processing events.
-		let _read_guard = $self.total_consistency_lock.read().unwrap();
+		let mut processed_all_events = false;
+		while !processed_all_events {
+			if $self.pending_events_processor.compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed).is_err() {
+				return;
+			}
 
-		let mut result = NotifyOption::SkipPersist;
+			let mut result = NotifyOption::SkipPersist;
 
-		// TODO: This behavior should be documented. It's unintuitive that we query
-		// ChannelMonitors when clearing other events.
-		if $self.process_pending_monitor_events() {
-			result = NotifyOption::DoPersist;
-		}
+			{
+				// We'll acquire our total consistency lock so that we can be sure no other
+				// persists happen while processing monitor events.
+				let _read_guard = $self.total_consistency_lock.read().unwrap();
+
+				// TODO: This behavior should be documented. It's unintuitive that we query
+				// ChannelMonitors when clearing other events.
+				if $self.process_pending_monitor_events() {
+					result = NotifyOption::DoPersist;
+				}
+			}
 
-		let pending_events = mem::replace(&mut *$self.pending_events.lock().unwrap(), vec![]);
-		if !pending_events.is_empty() {
-			result = NotifyOption::DoPersist;
-		}
+			let pending_events = $self.pending_events.lock().unwrap().clone();
+			let num_events = pending_events.len();
+			if !pending_events.is_empty() {
+				result = NotifyOption::DoPersist;
+			}
 
-		for event in pending_events {
-			$event_to_handle = event;
-			$handle_event;
-		}
+			for event in pending_events {
+				$event_to_handle = event;
+				$handle_event;
+			}
 
-		if result == NotifyOption::DoPersist {
-			$self.persistence_notifier.notify();
+			{
+				let mut pending_events = $self.pending_events.lock().unwrap();
+				pending_events.drain(..num_events);
+				processed_all_events = pending_events.is_empty();
+				$self.pending_events_processor.store(false, Ordering::Release);
+			}
+
+			if result == NotifyOption::DoPersist {
+				$self.persistence_notifier.notify();
```

Comment:
How often do we expect to loop multiple times? Just wonder if we'll be hitting this a lot and should persist more sparingly